### PR TITLE
TriangleSet optimizations - ~80% improvement in Ray picking against models

### PR DESCRIPTION
**Pull request by ZappoMan**
Implements ray picking performance improvements against models by adding a model frame octree to the TriangleSet class. Improves ray picking by ~80% by avoiding majority of triangles in the set.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: dcb70aa50467dbdf6264cb00f3f79dd7e8da0882 available at: https://gist.github.com/6babdf74f8a6b871d4b28e19ad67eff5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-dcb70aa.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-dcb70aa.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: da404ce2ce573358344e5efb253e8edcda1659ce available at: https://gist.github.com/6babdf74f8a6b871d4b28e19ad67eff5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-da404ce.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-da404ce.dmg

**Comment by ZappoMan** [comment]
# Test Plan:
- Smoke test focusing on "ray picking" - should be same behavior as master/beta. Verify that lasers terminate on models in the correct location. This PR will not change behavior of ray picking against primitives other than models, it only changes the code for how a ray pick is done against a model's mesh. Examples of things you can test ray picking with:
    - Teleport beam
    - mouse based clicking while in edit or with the mouse drag on physical objects
    - far grab
    - laser pointers

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5065c7c5c45712f058e5c7c3c762eab7f6bb9a1c available at: https://gist.github.com/6babdf74f8a6b871d4b28e19ad67eff5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-5065c7c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-5065c7c.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

all of the functions in the test plan appear to react the same as it does in the beta build, whilst performing the above in HMD and Desktop where applicable.

Let me know if there is more I can do/provide :)

**Comment by ZappoMan** [comment]
In the process of writing a performance test, I've found some unexplained cases where some ray picking seems to be failing. So for now, I'm going to say this is a fail until I know more.

**Comment by ZappoMan** [comment]
Ok - I figured it out... I'll have a new version up later tonight, with additional tests.

**Comment by ZappoMan** [comment]
This is ready to be re-tested

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9c81a89ac3e43b0adc671e3233593c233b2ddcb1 available at: https://gist.github.com/6babdf74f8a6b871d4b28e19ad67eff5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-9c81a89.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10455-9c81a89.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Having a few issues

The rayPickPerformance.js is not present in the master/beta build, but is accessible in this PR.

Also, I'm not seeing the red lines for the test appear at all, went out very far (1800 out) in the sandbox domain, figuring i'd have the required rights in that location.

Not sure where I'm going wrong. I'm still a bit new so I could be easily missing something?

**Comment by ZappoMan** [comment]
@YuppyPlays - you can also run the script from this URL - https://raw.githubusercontent.com/ZappoMan/hifi/9c81a89ac3e43b0adc671e3233593c233b2ddcb1/scripts/developer/tests/performance/rayPickPerformance.js

As for the red rays not rendering at all... I'm not sure why that would be... maybe your machine is struggling to render them, and the LOD is adjusting... maybe attach your log file so we can see what it says.

File: libraries/shared/src/TriangleSet.cpp [hunk 1]
```diff
@@ -56,21 +132,134 @@ bool TriangleSet::findRayIntersection(const glm::vec3& origin, const glm::vec3&
...
+            return false;
+        }
+
+        bestLocalDistance = distance;
+
+        float childDistance = distance;
+        BoxFace childFace;
+        glm::vec3 childNormal;
+
+        // if we're not yet at the max depth, then check which child the triangle fits in
```
Comment (AndrewMeadows) [inline libraries/shared/src/TriangleSet.cpp]: This comment doesn't make sense.  Maybe was a cut-n-paste from elsewhere?

**Comment by jherico** [comment]
The octree and balanced flag are purely internal state for the `TriangleSet`.  I would therefore suggest that you make them `mutable`, so that members of `TriangleSet` can manipulate them and still be declared `const`.  

I'd also love to see either manual or automated unit tests for the TriangleOctree, as it's a lot of new code to add without any tests.

File: libraries/shared/src/TriangleSet.h [hunk 2]
```diff
@@ -15,27 +15,75 @@
...
+
+    protected:
+        TriangleOctreeCell(std::vector<Triangle>& allTriangles, const AABox& bounds, int depth);
+
+        // checks our internal list of triangles
+        bool findRayIntersectionInternal(const glm::vec3& origin, const glm::vec3& direction,
+            float& distance, BoxFace& face, glm::vec3& surfaceNormal, bool precision, int& trianglesTouched);
+
+        std::vector<Triangle>& _allTriangles;
+        std::map<AABox::OctreeChild, TriangleOctreeCell> _children;
```
Comment (jherico) [inline libraries/shared/src/TriangleSet.h]: a `std::map` will sort the children, providing a LogN lookup time.  perhaps this should be `std::unordered_map`

Comment (ZappoMan) [inline libraries/shared/src/TriangleSet.h]: the max map size is 8 - seems unlikely to be a performance impact

**Comment by ZappoMan** [comment]
@jherico - I added a manual test in the form of `rayPickPerformance.js` is that not sufficient?

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Sorry this is taking so long, I was re-installing my High Fidelity versions and for some reason the latest beta is throwing errors at me regarding Can't Write data to C:/program files, etc...

Going to try a few things, will make a comment the moment I've finished the test.

**Comment by YuppyPlays** [comment]
Testing complete:

When running the latest beta build and using the provided .js script, the game crashed the first time, then the 2nd time just showed the red lines, no buddha (Picture below)

raytest


When running the provided PR with the provided script, the script works as expected. the logs show the following:

[05/16 23:07:52] [DEBUG] [hifi.scriptengine.script] ending loop:0
[05/16 23:07:52] [DEBUG] [hifi.scriptengine.script] beginning loop:1
[05/16 23:07:52] [DEBUG] [hifi.scriptengine.script] ending loop:1
[05/16 23:07:52] [DEBUG] [hifi.scriptengine.script] beginning loop:2
[05/16 23:07:53] [DEBUG] [hifi.scriptengine.script] ending loop:2
[05/16 23:07:53] [DEBUG] [hifi.scriptengine.script] beginning loop:3
[05/16 23:07:53] [DEBUG] [hifi.scriptengine.script] ending loop:3
[05/16 23:07:53] [DEBUG] [hifi.scriptengine.script] beginning loop:4
[05/16 23:07:54] [DEBUG] [hifi.scriptengine.script] ending loop:4
[05/16 23:07:54] [DEBUG] [hifi.scriptengine.script] beginning loop:5
[05/16 23:07:54] [DEBUG] [hifi.scriptengine.script] ending loop:5
[05/16 23:07:54] [DEBUG] [hifi.scriptengine.script] beginning loop:6
[05/16 23:07:55] [DEBUG] [hifi.scriptengine.script] ending loop:6
[05/16 23:07:55] [DEBUG] [hifi.scriptengine.script] beginning loop:7
[05/16 23:07:55] [DEBUG] [hifi.scriptengine.script] ending loop:7
[05/16 23:07:55] [DEBUG] [hifi.scriptengine.script] beginning loop:8
[05/16 23:07:56] [DEBUG] [hifi.scriptengine.script] ending loop:8
[05/16 23:07:56] [DEBUG] [hifi.scriptengine.script] beginning loop:9
[05/16 23:07:57] [DEBUG] [hifi.scriptengine.script] ending loop:9
[05/16 23:07:57] [DEBUG] [hifi.scriptengine.script] EXPECTED tests:13850 intersections:12860
[05/16 23:07:57] [DEBUG] [hifi.scriptengine.script] ACTUAL   tests:13850 intersections:12860
[05/16 23:07:57] [DEBUG] [hifi.scriptengine.script] ELAPSED TIME:5227 ms


Pass?

**Pull request merged**